pkg/sethandler/controller.go

```diff
@@ -113,35 +115,71 @@ func (setHandler *SetHandler) PodChanged(oldPod, newPod v1.Pod) {
 	if !shouldPodBeHandled(newPod) {
 		return
 	}
-	setHandler.adjustContainerSets(newPod)
+	if proceed, containersToBeSet := determineContainersToBeSet(oldPod, newPod); proceed {
+			setHandler.adjustContainerSets(newPod, containersToBeSet)
+	}
 }
 
 func shouldPodBeHandled(pod v1.Pod) bool {
 	setterNodeName := os.Getenv("NODE_NAME")
 	podNodeName := pod.Spec.NodeName
-
 	//Pod is not yet scheduled, or it wasn't scheduled to the Node of this specific CPUSetter instance
 	if setterNodeName == "" || podNodeName == "" || setterNodeName != podNodeName {
 		return false
 	}
-	//If the Pod is not running, its containers haven't been created yet - no cpuset cgroup is present to be adjusted by the CPUSetter
-	if pod.ObjectMeta.Annotations[setterAnnotationKey] != "" {
-		return false
-	}
-	if len(pod.Status.ContainerStatuses) == 0 {
-		return false
-	}
-	for _, containerStatus := range pod.Status.ContainerStatuses {
-		if containerStatus.ContainerID == "" {
-			return false
+	return true
+}
+
+func determineContainersToBeSet(oldPod, newPod v1.Pod)  (bool, []string){
+	workingContainers := []string{}
+	found := false
+	if newPod.ObjectMeta.Annotations[setterAnnotationKey] != "" && oldPod.ObjectMeta.Name != ""  {
+		if len(newPod.Status.ContainerStatuses) == 0 || len(oldPod.Status.ContainerStatuses) == 0 {
+			return false, []string{}
+		}
+		for _, newContainerStatus := range newPod.Status.ContainerStatuses {
+			found = false
+			for _, oldContainerStatus := range oldPod.Status.ContainerStatuses {
+				if oldContainerStatus.ContainerID == newContainerStatus.ContainerID {
+					found = true
+				}
+			}
+			if !found {
+				workingContainers = append(workingContainers, newContainerStatus.Name)
+			}
+		}
+	} else {
```

Comment:
and it would also mean we don't need this branch neither

Reply:
I rewrite it, but I think it's still needed. There is a scenario when podAdded failed/discarded, and next podChange still doesn't contain annotation, and need to add all container.

Reply:
yeah but that is what you can handle outside
you don't need to go in and check the statuse if each and every specific containers. no annotationa blindly equals creating a a set with all container IDs regardless if it is Add, or Update

Reply:
and if there is an annotation, now then you need to check if any of them are "new" ones

Reply:
I handle annotation check in PodChange, if annotation exists run determineContainersToBeSet, else run the same functionality as PodAdded.